### PR TITLE
CMake improvements

### PR DESCRIPTION
**Pull request by chfast**
- Allow setting custom installation prefix.
- Add option to disable building tests.

### CODE REVIEW

**Comment by chfast** [comment]
Anyone?

**Comment by chfast** [comment]
?

**Comment by yiwu-arbug** [comment]
@adsharma Do you have any comments?

File: CMakeLists.txt [hunk 1]
```diff
@@ -757,11 +757,11 @@ endforeach(sourcefile ${C_TEST_EXES})
 add_subdirectory(tools)
 
 # Installation and packaging for Linux
-if(NOT WIN32)
```
Comment (adsharma) [inline CMakeLists.txt]: Looks like you dropped the NOT WIN32 clause. Is it safe to do so on Windows?

File: CMakeLists.txt [hunk 2]
```diff
@@ -563,6 +563,12 @@ if(${CMAKE_INSTALL_PREFIX} STREQUAL /usr/local)
   set(CMAKE_INSTALL_PREFIX /usr)
 endif()
 
+option(WITH_TESTS "build with tests" ON)
+if(NOT WITH_TESTS)
```
Comment (adsharma) [inline CMakeLists.txt]: If someone adds code to the end of this file that is not related to tests, the return over here will break things in hard to debug ways.

I'd rather keep the testing code surrounded by WITH_TESTS instead of relying on the ordering.

File: CMakeLists.txt (hunk 1 above)
Comment (chfast) [inline CMakeLists.txt]: Yes. This is useful to be used together with ExternalProject.cmake module or Hunter package manager.

**Comment by chfast** [comment]
How it looks now?

**Comment by sagar0** [comment]
@chfast Thanks for contributing!
Seems like there are a few conflicts + all builds are failing. Can you rebase, and fix the issues.

File: CMakeLists.txt (hunk 2 above)
Comment (chfast) [inline CMakeLists.txt]: Done.

**Comment by chfast** [comment]
Can your review it now before I get another set of conflicts?

File: CMakeLists.txt [hunk 3]
```diff
@@ -810,14 +822,9 @@ foreach(sourcefile ${C_TEST_EXES})
     add_test(NAME ${exename} COMMAND ${exename}${ARTIFACT_SUFFIX})
     add_dependencies(check ${exename}${ARTIFACT_SUFFIX})
 endforeach(sourcefile ${C_TEST_EXES})
-add_subdirectory(tools)
+endif()
 
-# Installation and packaging for Linux
-if(NOT WIN32)
```
Comment (sagar0) [inline CMakeLists.txt]: Based on my understanding, it seems like installation and packaging are applicable only to non-win32 platforms so far. With the current change, CMake will attempt to install binaries to some path on windows as well. So, I don't quite understand how the check for WIN32 can be removed here. Could you please explain more about his change? (I know there there was an earlier question you addressed around this, but I still don't get it). 

Also does `include(GNUInstallDirs)` work on Windows? What values do `CMAKE_INSTALL_LIBDIR` and `CMAKE_INSTALL_INCLUDEDIR` get?

Comment (chfast) [inline CMakeLists.txt]: 1. Creating CMake packages (https://cmake.org/cmake/help/v3.8/manual/cmake-packages.7.html) is not for Unix-like systems only. See also ExternalPackage module and Hunter package manager.
2. The INSTALL step in CMake is required to create the CMake package because only the CMake files describing the package exported in the INSTALL step are relocatable and can be imported by other projects.
Maybe using `make install` equivalent on Windows to install files to `C:\Program Files` does not make much sense, but using this to build dependencies locally definitely does.
3. GNUInstallDirs has good defaults that work on Windows as well. Defaults are `bin`, `lib`, etc. GNUInstallDirs just solved issue on systems where libs should be installed to `lib64` instead of `lib` dir.

@ruslo, do you want to add something about CMake packages and Hunter?

Comment (ruslo) [inline CMakeLists.txt]: > do you want to add something about CMake packages and Hunter?

If this library works on Windows and somebody want to use it, then there should be install procedure available, it's pretty obvious :) This is true even out of scope of creating CMake package (you have to run install even if no `*Config.cmake` will be installed) or Hunter (there is no dependencies).

I see few general CMake issues, can comment on them if anybody interested.

Comment (chfast) [inline CMakeLists.txt]: I'm interested and happy to fix them step by step.

File: CMakeLists.txt [hunk 4]
```diff
@@ -600,6 +600,18 @@ else()
   message(STATUS "JNI library is disabled")
 endif()
 
+# Installation and packaging for Linux
+include(GNUInstallDirs)
+install(TARGETS ${ROCKSDB_STATIC_LIB} COMPONENT devel ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
+install(TARGETS ${ROCKSDB_SHARED_LIB} COMPONENT runtime DESTINATION ${CMAKE_INSTALL_LIBDIR})
```
Comment (ruslo) [inline CMakeLists.txt]: On Windows shared library produce two files DLL and LIB. Can be controlled by `ARCHIVE DESTINATION` and `RUNTIME DESTINATION`. Usually I'm using [this code](http://cgold.readthedocs.io/en/latest/tutorials/install/layout.html):
```cmake
install(
    TARGETS <project>_<target>_1 <project>_<target>_2
    EXPORT <project>Targets
    LIBRARY DESTINATION "lib"
    ARCHIVE DESTINATION "lib"
    RUNTIME DESTINATION "bin"
    INCLUDES DESTINATION "include"
)
```

Works for most cases if the target is not something exotic like OSX bundle/framework.

File: CMakeLists.txt [hunk 5]
```diff
@@ -600,6 +600,18 @@ else()
   message(STATUS "JNI library is disabled")
 endif()
 
+# Installation and packaging for Linux
+include(GNUInstallDirs)
+install(TARGETS ${ROCKSDB_STATIC_LIB} COMPONENT devel ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
```
Comment (ruslo) [inline CMakeLists.txt]: Installing simultaneously two versions (shared and static) of the same library in my opinion is a popular CMake antipattern. `add_library` + `BUILD_SHARED_LIBS=ON/OFF` should be used instead:
* http://cgold.readthedocs.io/en/latest/tutorials/libraries/static-shared.html

File: CMakeLists.txt [hunk 6]
```diff
@@ -600,6 +600,18 @@ else()
...
+install(TARGETS ${ROCKSDB_SHARED_LIB} COMPONENT runtime DESTINATION ${CMAKE_INSTALL_LIBDIR})
+install(DIRECTORY include/rocksdb COMPONENT devel DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
+if(${CMAKE_INSTALL_PREFIX} STREQUAL /usr/local)
```
Comment (ruslo) [inline CMakeLists.txt]: I'm as a user expect libraries to be installed to `/usr/local` when I do `cmake -DCMAKE_INSTALL_PREFIX=/usr/local/`. With this code behave is quite surprising for me - it will be installed to `/usr` instead even if I explicitly set another destination. 

In 99% cases the use of CMAKE_INSTALL_PREFIX in CMakeLists.txt means you're doing something wrong.

File: CMakeLists.txt (hunk 3 above)
Comment (chfast) [inline CMakeLists.txt]: @sagar0 ?

**Comment by sagar0** [comment]
@yuslepukhin Are you ok with me to accept and merge this PR? This looks ok to me, but wanted to get your confirmation as it especially affects Windows builds.

File: CMakeLists.txt (hunk 5 above)
Comment (yuslepukhin) [inline CMakeLists.txt]: +1

File: CMakeLists.txt (hunk 4 above)
Comment (yuslepukhin) [inline CMakeLists.txt]: Good point. We divorce names for static and dll.

File: CMakeLists.txt (hunk 3 above)
Comment (yuslepukhin) [inline CMakeLists.txt]: I am skeptical about creating a viable installation on windows. However, if you decided to go ahead I'd prefer to make it optional.

Comment (sagar0) [inline CMakeLists.txt]: @chfast Do you plan to make the installation optional on windows, as suggested above?

**Comment by sagar0** [comment]
For further community review, if any: @tamird @benesch

File: CMakeLists.txt (hunk 3 above)
Comment (chfast) [inline CMakeLists.txt]: Ok, but what the condition should be?

We cannot check nicely if the `CMAKE_INSTALL_PREFIX` was set by the user, because this var is never empty (has some OS-specific defaults).

How about `SNAPPY_INSTALL_ON_WINDOWS`?

@ruslo any ideas?

Comment (ruslo) [inline CMakeLists.txt]: > any ideas?

@chfast I don't see the point of making it conditional. How user should use libraries if there will be no way to install it? Though it doesn't hurt if it will be under some condition.

> How about SNAPPY_INSTALL_ON_WINDOWS?

Sounds good.

Comment (chfast) [inline CMakeLists.txt]: I've found a way to find out whenever CMAKE_INSTALL_PREFIX was set by the user -- CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT. But available only in CMake 3.6+.

I will go with SNAPPY_INSTALL_ON_WINDOWS.

Comment (chfast) [inline CMakeLists.txt]: I meant `ROCKDS_INSTALL_ON_WINDOWS`.

**Comment by chfast** [comment]
I rebased with hope to make a clean build on Windows, but it still failing by compiler running out of memory.

Anyway, any more comments here?

**Comment by sagar0** [comment]
Windows appveyor builds have been broken for a few days now. So, we can't test the current version on windows. But I know the earlier versions of this PR's diffs were successful on windows. 
So I am gonna go ahead and accept it in its current form. Thanks for taking everyone's suggestions into account.
